realtek: fix SMP startup (v2)

plappermaul opened the pull request:
The scope of the SMP startup structure is wrong. It is created on the stack and not as a global variable. This can lead to startup failures.

Fixes: 3f41360eb70c ("realtek: use upstream recommendation for CPU start")
Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de

plappermaul committed:
realtek: fix SMP startup

The scope of the SMP startup structure is wrong. It is created on the
stack and not as a global variable. This can lead to startup failures.

Fixes: 3f41360eb70c ("realtek: use upstream recommendation for CPU start")
Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de

musashino205 commented:
Is local scope not enough for ```vsmp_smp_ops```? It will only be used for copying to ```rtl_smp_ops``` .

plappermaul commented:
https://github.com/openwrt/openwrt/pull/10676 is the same deal.

plappermaul commented:
I just aligned.

musashino205 commented:
Oh, I see.

plappermaul commented:
@svanheule the second commit should explain your question about the timer driver. That coding has been dropped and was added here.

oliv3r commented on target/linux/realtek/files-5.10/arch/mips/rtl838x/prom.c:
```diff
@@ -35,11 +35,16 @@ static struct plat_smp_ops rtl_smp_ops;
 /*
- * MIPS timer interrupt might fire like crazy if not used or initialized
- * properly. Silence it by setting the maximum possible interval.
+ * These devices are low on resources so it is helpful to disable all unneeded
```
I think this first sentence about resources can be omitted, as eventually, we _might_ go for a single generic mips kernel for multiple devices, where CEVT_R4K might simply always be enabled.

oliv3r commented:
Reviewed-by: Olliver Schinagl <oliver@schinagl.nl>
Tested-by: Olliver Schinagl <oliver@schinagl.nl>
Acked-by: Olliver Schinagl <oliver@schinagl.nl>

on RTL9302 :)

svanheule commented:
@plappermaul Thanks for investigating! Very odd that the R4K timer doesn't generate interrupts on VPE0, but does generate them on VPE1... Now I'm also curious if this flag is already set on the other, non-triggering, VPE, and if the counter could instead be enable there.

svanheule commented on target/linux/realtek/files-5.10/arch/mips/rtl838x/prom.c:
```diff
@@ -35,11 +35,16 @@ static struct plat_smp_ops rtl_smp_ops;
+ * internal timer interrupt 7 that may be active by default. In this case
+ * disable counting in the core and confirm a pending interrupt.
  */
+	write_c0_cause(read_c0_cause() | BIT(27));
```
The described behaviour matches the definition of `BIT(27)` or `CAUSEF_DC` in the kernel's mipsregs.h. Please use the latter instead, since that's more descriptive. Not that I could find that much information, since it appears to be reserved on R4K.

plappermaul replied:
See https://s3-eu-west-1.amazonaws.com/downloads-mips/documents/MD00534-2B-34K-SUM-01.13.pdf page 203

plappermaul replied:
done

plappermaul committed:
realtek: SMP handling of R4K timer interrupts

Until now there has been no good explanation why we mess with the R4K
timer on SMP. After extensive testing and looking at the SDK code it
becomes clear what it is all about.

When we disable the CEVT_R4K module (we will do with the new timer
driver) the R4K timer hardware still fires interrupts on the secondary
CPU. To get around this we have two options:

- Disable IRQ 7
- Stop the counter completely

This patch selects option two because this is the root of evil.. To be
on the safe side we will do it only in case the CEVT_R4K module is
disabled.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

svanheule commented:
Thanks for the updates. Merged to master:
2b12da131370 realtek: fix SMP startup
b7aab1958591 realtek: SMP handling of R4K timer interrupts

svanheule commented:
```
ERROR: Unrecognized email address: 'Markus Stockhausen <markus.stockhausen@gmx.de'
#10: 
Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de

WARNING: Missing Signed-off-by: line by nominal patch author 'Markus Stockhausen <markus.stockhausen@gmx.de>'

total: 1 errors, 1 warnings, 15 lines checked

```

So, again, *please run checkpatch.pl on your own patches before submission*. The patches are already pushed out, so this can't be fixed any more.

svanheule commented:
And now I also see that you didn't include Olliver's tag in your latest revision. People put time and effort in reviewing and testing your patches, please include tags if they are provided.

svanheule commented:
Please compile test your changes too. Your last revision used a non-existent macro.

918e77465868 realtek: use correct CAUSEF_DC macro in prom.c

plappermaul commented:
You are pushing a v2 PR of me to master? Feels like birthday and christmas on one day. Fun aside. Completely understood. And no problem to apologize for the inconvenience and to thank for your fix. I can understand how you feel. Regarding the time spent one should not forget the multiples for gettting that developed and pre-testing this on three different devices.

The epic challenge between maintainer and developer ...

oliv3r commented:
> And now I also see that you didn't include Olliver's tag in your latest revision. People put time and effort in reviewing and testing your patches, please include tags if they are provided.

Heh, I don't mind; not sure what the OpenWRT process is; I think the author is to add tags when doing v(n) -> v(n+1), otherwise the maintainer is expected to add them. But we are talking ML stuff here, on github; you probably can/need-to? Edit a MR? Idk :shrug:
